File: node_backend/src/dbops/user_dbops.ts
// dbops/user_dbops.ts — Equivalent to go_backend/dbops/user_dbops.go
import { PrismaClient } from '@prisma/client';
import { User } from '../entities/user';

/**
 * Interface for User database operations.
 * Mirrors the Go UserDBOps interface for dependency injection.
 */
export interface UserDBOps {
    create(data: { name: string; email: string }): Promise<User>;
    getByID(id: number): Promise<User | null>;
    getByEmail(email: string): Promise<User | null>;
}

/**
 * Prisma-based implementation of UserDBOps.
 * Equivalent to the Go userDBOps struct.
 */
export class PrismaUserDBOps implements UserDBOps {
    private db: PrismaClient;

    constructor(db: PrismaClient) {
        this.db = db;
    }

    async create(data: { name: string; email: string }): Promise<User> {
        return this.db.user.create({
            data: {
                name: data.name,
                email: data.email,
            },
        });
    }

    async getByID(id: number): Promise<User | null> {
        return this.db.user.findFirst({
            where: {
                id,
                deletedAt: null, // Soft delete filter
            },
        });
    }

    async getByEmail(email: string): Promise<User | null> {
        return this.db.user.findFirst({
            where: {
                email,
                deletedAt: null, // Soft delete filter
            },
        });
    }
}

File: go_backend/dbops/user_dbops.go
package dbops

import (
	"context"

	"syncflow/entities"
	"gorm.io/gorm"
)

type UserDBOps interface {
	Create(ctx context.Context, user *entities.User) error
	GetByID(ctx context.Context, id uint) (*entities.User, error)
	GetByEmail(ctx context.Context, email string) (*entities.User, error)
}

type userDBOps struct {
	db *gorm.DB
}

func NewUserDBOps(db *gorm.DB) UserDBOps {
	return &userDBOps{db: db}
}

func (d *userDBOps) Create(ctx context.Context, user *entities.User) error {
	return d.db.WithContext(ctx).Create(user).Error
}

func (d *userDBOps) GetByID(ctx context.Context, id uint) (*entities.User, error) {
	var user entities.User
	err := d.db.WithContext(ctx).First(&user, id).Error
	if err != nil {
		return nil, err
	}
	return &user, nil
}

func (d *userDBOps) GetByEmail(ctx context.Context, email string) (*entities.User, error) {
	var user entities.User
	err := d.db.WithContext(ctx).Where("email = ?", email).First(&user).Error
	if err != nil {
		return nil, err
	}
	return &user, nil
}
